DFC.ServiceTaxonomy.GraphSync/GraphSyncers/Parts/FlowPartSyncer.cs

```diff
@@ -0,0 +1,257 @@
+            IGraphSyncHelper graphSyncHelper,
+            IGraphValidationHelper graphValidationHelper,
+            IDictionary<string, int> expectedRelationshipCounts,
+            string endpoint)
+        {
+            (bool validated, string failureReason) = await ValidateWidgetsSyncComponent(
+                content,
+                nodeWithOutgoingRelationships,
+                graphValidationHelper,
+                expectedRelationshipCounts,
+                endpoint);
+
+            if (!validated)
+                return (validated, failureReason);
+
+            using var _ = graphSyncHelper.PushPropertyNameTransform(_flowFieldsPropertyNameTransform);
+
+            (validated, failureReason) = await _contentFieldsGraphSyncer.ValidateSyncComponent(
+                content,
+                contentTypePartDefinition,
+                nodeWithOutgoingRelationships,
+                graphSyncHelper,
+                graphValidationHelper,
+                expectedRelationshipCounts);
+
+            return (validated, failureReason);
+        }
+
+        private async Task<(bool validated, string failureReason)> ValidateWidgetsSyncComponent(
+            JObject content,
+            INodeWithOutgoingRelationships nodeWithOutgoingRelationships,
+            IGraphValidationHelper graphValidationHelper,
+            IDictionary<string, int> expectedRelationshipCounts,
+            string endpoint)
+        {
+            IEnumerable<ContentItem>? contentItems = content[Widgets]?.ToObject<IEnumerable<ContentItem>>();
+            if (contentItems == null)
+                throw new GraphSyncException("Flow does not contain Widgets");
+
+            int widgetOrdinal = 0;
+            foreach (ContentItem flowPartContentItem in contentItems)
+            {
+                var graphSyncValidator = _serviceProvider.GetRequiredService<IValidateAndRepairGraph>();
+
+                ContentTypeDefinition flowPartContentTypeDefinition = _contentTypes[flowPartContentItem.ContentType];
+
+                (bool validated, string failureReason) =
+                    await graphSyncValidator.ValidateContentItem(flowPartContentItem, flowPartContentTypeDefinition,
+                        endpoint);
+
+                if (!validated)
+                    return (false, $"contained item failed validation: {failureReason}");
+
+                // check expected relationship is in graph
+                var flowContentGraphSyncHelper = _serviceProvider.GetRequiredService<IGraphSyncHelper>();
+
+                flowContentGraphSyncHelper.ContentType = flowPartContentItem.ContentType;
+                string expectedRelationshipType = await RelationshipType(flowContentGraphSyncHelper);
+
+                // keep a count of how many relationships of a type we expect to be in the graph
+                expectedRelationshipCounts.TryGetValue(expectedRelationshipType, out int currentCount);
```

Comment:
Can't see the full stack here but is there a possibility that TryGetValue will not find the expectedRelationshipType?

Reply:
Yes it won't find the expected relationship type the first time it encounters that type, but then currentCount will be 0, and the next line adds the expected relationship type with a count of 1.
So basically, not finding it is part of the normal flow and is handled.